File: tests/fixed-stack.js
const { native, addIterableMethods } = require('../adapter/_helpers.js');
const protocol = process.env.MNEMONIST_TRANSPORT === 'protocol'
  ? require('../adapter/_protocol.js')
  : null;

function FixedStack(ArrayClass, capacity) {
  if (arguments.length < 2) {
    throw new Error('mnemonist/fixed-stack: expecting an Array class and a capacity.');
  }
  if (typeof capacity !== 'number' || capacity <= 0) {
    throw new Error('mnemonist/fixed-stack: `capacity` should be a positive number.');
  }
  this._inner = protocol
    ? protocol.create('fixed-stack', [capacity])
    : new native.FixedStackInner(capacity);
  this.capacity = capacity;
  this.ArrayClass = ArrayClass;
}

FixedStack.prototype.clear = function() { return protocol ? this._inner.call('clear', []) : this._inner.clear(); };
FixedStack.prototype.push = function(item) { return protocol ? this._inner.call('push', [item]) : this._inner.push(item); };
FixedStack.prototype.pop = function() { return protocol ? this._inner.call('pop', []) : this._inner.size === 0 ? undefined : this._inner.pop(); };
FixedStack.prototype.peek = function() { return protocol ? this._inner.call('peek', []) : this._inner.size === 0 ? undefined : this._inner.peek(); };
FixedStack.prototype.toArray = function() {
  const arr = protocol ? this._inner.call('toArray', []) : this._inner.toArray();
  if (this.ArrayClass && this.ArrayClass !== Array) {
    const typed = new this.ArrayClass(arr.length);
    for (let i = 0; i < arr.length; i++) typed[i] = arr[i];
    return typed;
  }
  return arr;
};

Object.defineProperty(FixedStack.prototype, 'size', {
  get: function() { return protocol ? this._inner.size() : this._inner.size; }
});

addIterableMethods(FixedStack.prototype, function() {
  const arr = this.toArray();
  return Array.from(arr);
});

FixedStack.from = function(iterable, ArrayClass, capacity) {
  if (arguments.length < 3) {
    capacity = iterable.length;
    if (typeof capacity !== 'number') {
      throw new Error('mnemonist/fixed-stack.from: could not guess iterable length. Please provide desired capacity as last argument.');
    }
  }
  const stack = new FixedStack(ArrayClass, capacity);
  if (typeof iterable.length === 'number') {
    for (let i = 0; i < iterable.length; i++) stack.push(iterable[i]);
    return stack;
  }
  for (const value of iterable) stack.push(value);
  return stack;
};

module.exports = FixedStack;

File: src/fixed_stack.rs
use serde_json::Value;

#[derive(Debug, Clone)]
pub struct FixedStack {
    capacity: usize,
    items: Vec<Option<Value>>,
    size: usize,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FixedStackError(pub String);

impl std::fmt::Display for FixedStackError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl std::error::Error for FixedStackError {}

impl FixedStack {
    pub fn new(capacity: usize) -> Result<Self, FixedStackError> {
        if capacity == 0 {
            return Err(FixedStackError(
                "mnemonist/fixed-stack: `capacity` should be a positive number.".into(),
            ));
        }

        Ok(Self {
            capacity,
            items: vec![None; capacity],
            size: 0,
        })
    }

    pub fn capacity(&self) -> usize {
        self.capacity
    }

    pub fn size(&self) -> usize {
        self.size
    }

    pub fn clear(&mut self) {
        self.size = 0;
    }

    pub fn push(&mut self, item: Value) -> Result<usize, FixedStackError> {
        if self.size == self.capacity {
            return Err(FixedStackError(format!(
                "mnemonist/fixed-stack.push: stack capacity ({}) exceeded!",
                self.capacity
            )));
        }

        self.items[self.size] = Some(item);
        self.size += 1;
        Ok(self.size)
    }

    pub fn pop(&mut self) -> Option<Value> {
        if self.size == 0 {
            return None;
        }
        self.size -= 1;
        self.items[self.size].take()
    }

    pub fn peek(&self) -> Option<&Value> {
        if self.size == 0 {
            return None;
        }
        self.items[self.size - 1].as_ref()
    }

    pub fn for_each<F>(&self, mut callback: F)
    where
        F: FnMut(&Value, usize),
    {
        let null = Value::Null;
        let l = self.capacity;
        for i in 0..l {
            let idx = l - i - 1;
            let item = self.items[idx].as_ref().unwrap_or(&null);
            callback(item, i);
        }
    }

    pub fn to_array(&self) -> Vec<Value> {
        let mut array = Vec::with_capacity(self.size);
        let mut i = self.size;
        while i > 0 {
            i -= 1;
            if let Some(ref v) = self.items[i] {
                array.push(v.clone());
            }
        }
        array
    }

    pub fn values(&self) -> impl Iterator<Item = &Value> {
        (0..self.size).rev().filter_map(|i| self.items[i].as_ref())
    }

    pub fn entries(&self) -> impl Iterator<Item = (usize, &Value)> {
        (0..self.size)
            .rev()
            .enumerate()
            .filter_map(|(i, idx)| self.items[idx].as_ref().map(|v| (i, v)))
    }

    pub fn from_iter<I>(iterable: I, capacity: Option<usize>) -> Result<Self, FixedStackError>
    where
        I: IntoIterator<Item = Value>,
        I::IntoIter: ExactSizeIterator,
    {
        let iter = iterable.into_iter();
        let cap = capacity.unwrap_or_else(|| iter.len());

        if cap == 0 {
            return Err(FixedStackError(
                "mnemonist/fixed-stack.from: could not guess iterable length. Please provide desired capacity as last argument.".into(),
            ));
        }

        let mut stack = Self::new(cap)?;
        for (i, value) in iter.enumerate() {
            stack.items[i] = Some(value);
            stack.size = i + 1;
        }
        Ok(stack)
    }
}
